app/models/interactive_run_state.rb

```diff
@@ -114,18 +114,39 @@ def has_linked_interactive
     (interactive.respond_to? :linked_interactive) && !interactive.linked_interactive.nil?
   end
 
-  # this seems like it should be modeled as a through association, but perhaps it is hard because
-  # interactives are polymorphic
-  def find_linked_interactive_state
-    return nil unless has_linked_interactive
-    linked_interactive = interactive.linked_interactive
-    runs = run.sequence_run ? run.sequence_run.runs : [run]
-    InteractiveRunState.where(run_id: runs.map(&:id), interactive_id: linked_interactive.id).first
+  # Returns all linked states. 1st element in the array is an interactive directly linked to given one.
+  def all_linked_states
+    result = []
+    current_interactive = interactive
+    while (current_interactive.respond_to? :linked_interactive) && !current_interactive.linked_interactive.nil?
+      linked_interactive = current_interactive.linked_interactive
+      runs = run.sequence_run ? run.sequence_run.runs : [run]
+      state = InteractiveRunState.where(run_id: runs.map(&:id), interactive_id: linked_interactive.id).first
+      # It's not necessary to provide some metadata even for empty linked states, although it doesn't cost much
+      # and it can be useful for interactive itself to know that there are some linked interactives,
+      # but not run by student yet.
+      linked_state_info = { interactive_id: current_interactive.id, data: nil, created_at: nil, updated_at: nil }
```

Comment:
Should the interactive_id be the `linked_interactive.id` instead?
If that is true, then it might be clearer to create a new method on InteractiveRunState that just returns the state_info hash.

Reply:
Hm, not sure if I understand. Do you mean just the key name (`interactive_id` vs `linked_interactive_id`) or some logic (since you used `.id` here)?
Actually, this ID not necessary and used by anything, but I just added it so there's any form of some unique ID provided in case the interactive needs to work more with those states.